dbms/src/DataStreams/RuntimeFilter.cpp

```diff
@@ -0,0 +1,195 @@
+    {
+        return;
+    }
+    Block data(std::initializer_list<ColumnWithTypeAndName>({values}));
+    switch (rf_type)
+    {
+    case tipb::IN:
+        try
+        {
+            LOG_DEBUG(log, "update values in rf, data size:{}", data.rows());
+            in_values_set->insertFromBlock(data, false, true);
+        }
+        catch (Exception & e)
+        {
+            if (e.code() == ErrorCodes::SET_SIZE_LIMIT_EXCEEDED)
+            {
+                std::string tmp_err_msg = "The rf in values exceed the limit. " + e.message();
+                updateStatus(RuntimeFilterStatus::FAILED, tmp_err_msg);
+                LOG_WARNING(log, "cancel runtime filter id:{}, reason: {} ", id, tmp_err_msg);
+            }
+        }
+        break;
+    case tipb::MIN_MAX:
+    case tipb::BLOOM_FILTER:
+        // todo
+        break;
+    }
+}
+
+void RuntimeFilter::finalize(const LoggerPtr & log)
+{
+    if (isFailed() || isReady())
+    {
+        return;
+    }
+    updateStatus(RuntimeFilterStatus::READY);
+    std::string rf_values_info;
+    switch (rf_type)
+    {
+    case tipb::IN:
+        rf_values_info = fmt::format("number of IN values:{}", in_values_set->getTotalRowCount());
+        break;
+    case tipb::MIN_MAX:
+    case tipb::BLOOM_FILTER:
+        // TODO
+        break;
+    }
+    LOG_INFO(log, "finalize runtime filter id:{}, rf values info:{}", id, rf_values_info);
+}
+
+void RuntimeFilter::cancel(const LoggerPtr & log, const std::string & reason)
+{
+    if (isFailed() || isReady())
+    {
+        return;
+    }
+    updateStatus(RuntimeFilterStatus::FAILED, reason);
+    LOG_INFO(log, "cancel runtime filter id:{}, reason:{}", id, reason);
+}
+
+bool RuntimeFilter::isReady()
+{
+    return status == RuntimeFilterStatus::READY;
+}
+
+bool RuntimeFilter::isFailed()
+{
+    return status == RuntimeFilterStatus::FAILED;
+}
+
+bool RuntimeFilter::await(int64_t ms_remaining)
+{
+    if (isFailed())
+    {
+        return false;
+    }
+    if (!isReady())
+    {
+        if (ms_remaining <= 0)
+        {
+            return isReady();
+        }
+        std::unique_lock<std::mutex> lock(inner_mutex);
+        return inner_cv.wait_for(lock, std::chrono::milliseconds(ms_remaining), [this] { return isReady(); });
+    }
+    return true;
+}
+
+void RuntimeFilter::updateStatus(RuntimeFilterStatus status_, const std::string & reason)
+{
+    switch (status_)
+    {
+    case RuntimeFilterStatus::NOT_READY:
+        // The initial state of the runtime filter is not ready. It can only be updated as FAILED or READY.
+        return;
+    case RuntimeFilterStatus::FAILED:
+    {
+        {
+            std::lock_guard<std::mutex> lock(mtx);
+            if (status == RuntimeFilterStatus::READY)
+            {
+                return;
+            }
+            if (failed_reason.empty())
+            {
+                failed_reason = reason;
+            }
+            break;
+        }
+    }
+    case RuntimeFilterStatus::READY:
```

Comment:
The status still can be changed to `READY` from `FAILED`? And I think we can only allow valid status transition inside this `updateStatus` function, and when status is changed return `true`, otherwise return `false`, then we don't need to care about the current status before calling `updateStatus`(currently, current status is checked in `RuntimeFilter::cancel` and `RuntimeFilter::finalize`)

Reply:
No, the status can't change from ```READY``` to ```FAILED```.
If runtime filter status is ```READY```, then it can not be changed to ```FAILED```. The ```cancel ``` function also cannot change a ```READY``` runtime filter

Reply:
I return the bool if status has been updated